Add Lens Distortion Estimator plugin

10110111 opened the pull request:
This is a new plugin that can be used to estimate distortion of the objective lens of your camera by taking a photo of stars, picking them (by hand) to match to Stellarium's simulation, and generate a distortion model for use with the lensfun library. How to use it is written in the About tab of the config dialog.

It currently works (only rectilinear lenses are supported for now), but I'm not completely satisfied with the UI, particularly for initial positioning of the image in the sky, and also for picking the points. Maybe you can come up with suggestions on how to make it easier to use (maybe some dragging to move and rotate the image? is there any object with such capabilities in Stellarium?).

If the instructions in the About text are unclear, I'd also like to get your feedback.

Another incompleteness is that the core dependency, GSL, is not downloaded in any way, only located via `find_package()`. If it's missing, you can only fit manually, which is very tedious and generally unwanted since you can simply build with GSL support and click the semi-magic button.

One more dependency, exiv2, is not so much required, but it lets you easily set simulation time to the photo time from EXIF tags. This dependency is also only optionally used.

I'm not sure what we want to do with these dependencies: download from the releases? For exiv2 it seems easy, but GSL doesn't release binaries — should we build them on each rebuild of Stellarium, or maybe host the binaries we build once? Or maybe use CPM for both (which would add more time to the build process)?

gzotti commented:
Cool!

Some first observations: The spinners could be modified with AngleSpinBoxes. This would allow mouse-wheeling in units depending on cursor position. 

I just built on Windows, so probably I am missing most of the fun with GSL and exiv2. On image loading, I expect it to  be placed in the current screen center. Probably without exiv2 the fov was wrong. Some super-easy adjustment mode with e.g. ctrl+mousemove and ctrl+mousewheel could allow dragging and zooming the image. OTOH, Ctrl-drag is time-dragging, and Ctrl-mousewheel also shifts time, so maybe some other key, or configurable action keys (default Ctrl-shift-wheel?)  

Clicking on the image target pixel seems to not hit the exactly clicked pixel. I clicked in the center of that white blob...
![grafik](https://github.com/Stellarium/stellarium/assets/18099873/e62f0967-3dfc-4112-a007-26851d3b1fc3)


I don't know your motivation of finding lens distortion for single photos, but stars (high above the horizon) are certainly appropriate for this. Stitching panos with Hugin does that as side effect. In what workflow would I generate/export/copy-paste the XML? 

A similar tool would be great to interactively adjust constellation artwork. Speaking of that, using more than 3 link points (with some more work, triangulation or rubberbanding?) would also be highly welcome.

10110111 commented:
> On image loading, I expect it to be placed in the current screen center.

Regardless of where you are looking? Well, OK, this may be sensible, if you think the user would direct the view to where the camera looked before loading the image.

> Probably without exiv2 the fov was wrong.

Not really, although EXIF may contain focal length, it usually gives no info about sensor dimensions, so I can't compute FoV, just use a default instead. Well, maybe some manufacturers place some custom tags for this, this I need to research.

> maybe some other key, or configurable action keys (default Ctrl-shift-wheel?)

Let's say <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+wheel for FoV, and then we also need to:
* Move the image around
* Rotate it around its center

While moving around might be achieved with <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+move (removing the side effect of creating a marker, which should only happen on <kbd>Shift</kbd>), how to control rotation? Is there anything existing that would serve as an example?

> Clicking on the image target pixel seems to not hit the exactly clicked pixel.

OK, this I also noticed...

> I don't know your motivation of finding lens distortion for single photos, <...>. Stitching panos with Hugin does that as side effect.

It's not **for** single photos, but rather **by** a single photo per camera+lens. After this characterization you can fill in the lensfun database with your data and correct all subsequent photos with its tools or with RAW development programs.

Dissatisfaction with the results given out by Hugin and then the discovery that my Galaxy S8 specifies focal length that doesn't combine well with the sensor dimensions I find in specs (and in lensfun database!) was the main driving force that made me start this project. And after implementing this I noticed that lensfun database has zeros (almost?) everywhere for center positions, while in two actual devices (Galaxy S8 and a Canon DSLR) I found tens of pixels of shift from the center (i.e. the lens is not centered perfectly). I suppose the shift will vary from camera to camera, so it's useful for users to be able to easily characterize their own devices without the need to fiddle around large photosets suitable for creation of a panorama, especially since it has quite low success rate in my experience.

> In what workflow would I generate/export/copy-paste the XML?

Well, from [this tutorial](https://hugin.sourceforge.io/tutorials/calibration/en.shtml) I see that Hugin can save lens/camera parameters to lensfun database (directly?). Didn't try it though. Maybe it's better than presenting an XML snippet.

> A similar tool would be great to interactively adjust constellation artwork. Speaking of that, using more than 3 link points (with some more work, triangulation or rubberbanding?) would also be highly welcome.

Not sure what link points you mean. This plugin uses any number of points you pick ≥2, and, if built with GSL, it can optimize the RMS error between the points in the image and the simulated stars.

gzotti commented:
> > On image loading, I expect it to be placed in the current screen center.
> 
> Regardless of where you are looking? Well, OK, this may be sensible, if you think the user would direct the view to where the camera looked before loading the image.

Yes, I would start with recreating the view in Stellarium (with the photo in some image viewer) before importing the photo. Maybe use 80% vertical fov to apply a first scaling. Also, most images will be either horizontal or equatorial. Just take current "mount mode" into account to show the image horizontally on screen at import. This should then provide a "rough first match". 


> > Probably without exiv2 the fov was wrong.
> 
> Not really, although EXIF may contain focal length, it usually gives no info about sensor dimensions, so I can't compute FoV, just use a default instead. Well, maybe some manufacturers place some custom tags for this, this I need to research.

Ah right. And some manual lenses don't even report focal length. Maybe "grab" the instrument list from the Oculars plugin to allow selecting instrumentation (camera for sensor size, telescope+lens for focal length) before import. 
My (Canon) CR2 image has FocalPlaneXRresolution/FocalPlaneYResolution tags, giving pixels per inch (I think). Together with ExifImageWidth, ExifImageHeight this should allow estimating crop factor. But it seems the JPG loses those. 


> > maybe some other key, or configurable action keys (default Ctrl-shift-wheel?)
> 
> Let's say Ctrl+Shift+wheel for FoV, and then we also need to:
> 
>     * Move the image around
> 
>     * Rotate it around its center
> 
> 
> While moving around might be achieved with Ctrl+Shift+move (removing the side effect of creating a marker, which should only happen on Shift), how to control rotation? Is there anything existing that would serve as an example?

I am afraid no. Ctrl-Shift-Alt-wheel? 


> > Clicking on the image target pixel seems to not hit the exactly clicked pixel.
> 
> OK, this I also noticed...

May be related to Aberration. I have not tried to switch that off for a test. 

> 
> > I don't know your motivation of finding lens distortion for single photos, <...>. Stitching panos with Hugin does that as side effect.
> 
> It's not **for** single photos, but rather **by** a single photo per camera+lens. After this characterization you can fill in the lensfun database with your data and correct all subsequent photos with its tools or with RAW development programs.
> 
> Dissatisfaction with the results given out by Hugin and then the discovery that my Galaxy S8 specifies focal length that doesn't combine well with the sensor dimensions I find in specs (and in lensfun database!) was the main driving force that made me start this project. And after implementing this I noticed that lensfun database has zeros (almost?) everywhere for center positions, while in two actual devices (Galaxy S8 and a Canon DSLR) I found tens of pixels of shift from the center (i.e. the lens is not centered perfectly). I suppose the shift will vary from camera to camera, so it's useful for users to be able to easily characterize their own devices without the need to fiddle around large photosets suitable for creation of a panorama (or to search for very straight objects to picture), especially since it has quite low success rate in my experience.

Ah OK. This is for pretty advanced photographers then. I know one or two who would probably use it. 

> > In what workflow would I generate/export/copy-paste the XML?
> 
> Well, from [this tutorial](https://hugin.sourceforge.io/tutorials/calibration/en.shtml) I see that Hugin can save lens/camera parameters to lensfun database (directly?). Didn't try it though. Maybe it's better than presenting an XML snippet.

If there is use beyond Hugin, you don't have to interface with it. I just wondered about what this is all used for in practice. I have read about lensfun, but am not overly familiar. 
 
> > A similar tool would be great to interactively adjust constellation artwork. Speaking of that, using more than 3 link points (with some more work, triangulation or rubberbanding?) would also be highly welcome.
> 
> Not sure what link points you mean. This plugin uses any number of points you pick ≥2, and, if built with GSL, it can optimize the RMS error between the points in the image and the simulated stars.

Constellation artwork is linked with three stars and three pixel locations. The image is then stretched as required. When projections mismatch, the figure may become distorted. Having more link points could improve the figure match e.g. for old copperprint depictions. I would have to see the effect of GLS. Even a 3-point matching in such visual tool would be a great help. The end result should be lines compatible with, or even an interactive edit tool, for constellationart.fab files.  This may however become yet another plugin based on this one (It would have no camera specifics or color shift options). 

Of course, also self-made DSO figures could be adjusted with this tool. I have only ever added one DSO image (slide scan of &eta; Car) many years ago, section 8.2 of the User Guide has a current workflow. Looks a bit intimidating for a starter or casual user. The result box in this workflow could be a JSON snippet like in Figure 8.7.

10110111 commented:
> Ah OK. This is for pretty advanced photographers then.

Well, this depends on your definition of advanced :) My expectation is that this would be useful for anyone who wants to characterize their camera+lens devices for creation of accurate panoramas (I want them to have a reliable horizon!), or even simply for showing their scenery overlain with Stellarium's markers/labels/etc. So anyone who understands how to turn off geometry correction in their camera or how to develop RAW shots is advanced enough IMO. And yes, they have Hugin's optimizer as an alternative, it just imposes another set of requirements on the input photos.

> If there is use beyond Hugin, you don't have to interface with it.

I din't mean interfacing with Hugin, just borrowing the ideas for the UI. There's a lot of use of lensfun beyond Hugin: e.g. RawTherapee and Darktable both use lensfun for geometry/TCA/vignetting correction.

gzotti commented:
> > Ah OK. This is for pretty advanced photographers then.
> 
> Well, this depends on your definition of advanced :) My expectation is that this would be useful for anyone who wants to characterize their camera+lens devices for creation of accurate panoramas (I want them to have a reliable horizon!), or even simply for showing their scenery overlain with Stellarium's markers/labels/etc. So anyone who understands how to turn off geometry correction in their camera or how to develop RAW shots is advanced enough IMO. And yes, they have Hugin's optimizer as an alternative, it just imposes another set of requirements on the input photos.

I always had assumed Hugin's optimisation (gradually down to the largest set of variables, if required including shifts) would create a sufficient match with good input images. I have 30 full examples with measured polygons to match against, and matched accurate to 1-2 arcminutes. The quality of input images and keeping camera location where it should be is important of course.  (In landscape panos with just grassland foreground I don't need a tripod, though.) Also, switching off lens stabilizers (which may shift lens geometry) is a requirement for accurate results. If I can input known lens characteristics before starting Hugin's optimizer, this would be added value. Please also describe such workflows in the User Guide, chapter 13, and/or extend chapter 7 with improved workflow. 

> 
> > If there is use beyond Hugin, you don't have to interface with it.
> 
> I din't mean interfacing with Hugin, just borrowing the ideas for the UI. There's a lot of use of lensfun beyond Hugin: e.g. RawTherapee and Darktable both use lensfun for geometry/TCA/vignetting correction.

Ah, OK. I have no time to use any of those atm, but yes, fine!

10110111 commented:
I've just noticed that some stars don't have even an English name, not only localized one. Is there any handle except the pointer for such stars — e.g. so that exporting as a string and re-importing could work?

gzotti commented:
Brighter stars (Star1 class) have HIPS numbers. The dimmer ones report an empty name. These could only be identified via J2000 coordinates. (See also telescopic asterisms with dim stars.)

10110111 commented:
> These could only be identified via J2000 coordinates.

And to what precision should I serialize them to make deserialization unambiguous? Are these coordinates computed or simply loaded so that I can just output full `double` precision?

gzotti commented:
I also must study the code for that. See StarWrapper::getJ2000EquatorialPos() and StarMgr::searchAround(). The latter identifies the closest stars within a settable margin. It likely even depends on your image fov now. Arcminutes for wide-angle photos, but probably arcseconds for telescope shots.

10110111 commented:
What do the properties like X0 and X1 mean in the Star{1,2,3} structures?

gzotti commented:
Please see User Guide, Appendix C, table C.4. No full coordinates, just offsets from zone center. IMHO you should consider them private. At some point this star catalog format will have to change. (At least the data need a thorough inspection.)

10110111 commented:
Sorry for the huge amount of spam from the CI, I'm fighting its stubbornness...

10110111 commented:
So, now the new version with NLopt instead of GSL successfully builds, and you don't have to do anything special on Windows to get this functionality, thanks to CPM. I'll later remove the test for `HAVE_NLOPT`, since it should be always available.

As for exiv2, it seems unsuitable for CPM due to additional build dependencies. Not sure how I'll solve this.

Most of the functionality is now present, I've addressed most of the comments here. The image can be dragged to move and rotate, the initial placement is in the center of the screen, the initial vertical FoV is 98% of screen height, the wrong positioning on click is hopefully fixed.

As for the model for lensfun, it appears that Hugin's save to file and to database options only save the model to Hugin's custom INI format, or to its custom SQLite database, not to lensfun's database, and not in lensfun's format. So I suppose an XML output for pasting into lensfun's DB will suffice for now, until someone comes up with ideas for a better UX.

gzotti commented:
Great! I cannot test likely before Saturday, but anyone else please?

alex-w commented:
Great plugin, thanks!

> So I suppose an XML output for pasting into lensfun's DB will suffice for now, until someone comes up with ideas for a better UX.

+1 for XML

alex-w commented:
Is it still draft?

10110111 commented:
Yes, as I said before, I haven't found a good way to include exiv2 in the Windows builds (and we now have three architecture, not just two as before). If you have an idea, it will be appreciated.

10110111 commented:
> I cannot test likely before Saturday, but anyone else please?

Have you had an opportunity to try this PR?

gzotti commented:
I had read it's still a draft because of other issues. Sorry, no, had a repair in the house instead, and must reconsider things for a public presentation next week. (Must try out 23.1 instead...)

10110111 commented:
So finally I've decided to include exiv2 only in x86_64 Windows builds for now (just using the release binary), following the path of least resistance. Maybe later I'll be able to build it for other platforms, but for now I suppose this is sufficient.

Linux and macOS just use the versions installed during the CI. I'm not sure though whether the CI should pack the binary into macOS release somehow. It builds a custom version during installation, so the resulting binary needs to get into the release I guess.

alex-w commented:
> I'm not sure though whether the CI should pack the binary into macOS release somehow. It builds a custom version during installation, so the resulting binary needs to get into the release I guess.

I can share full log for building macOS package

10110111 commented:
> I can share full log for building macOS package

Log?.. I'd rather see (and edit) a script. Is it in the repo?

alex-w commented:
> > I can share full log for building macOS package
> 
> Log?.. I'd rather see (and edit) a script. Is it in the repo?

Hmm... see `CMakeLists.txt` and `src/CMakeLists.txt`

```
-- Performing Test HAS_FPIC - Success
CMake Warning (dev) at builds/macosx/_deps/nlopt-src/CMakeLists.txt:299 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonInterp: /opt/homebrew/bin/python3.10 (found version "3.10.13") 
CMake Warning (dev) at builds/macosx/_deps/nlopt-src/CMakeLists.txt:300 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonLibs: /opt/homebrew/Frameworks/Python.framework/Versions/3.10/lib/libpython3.10.dylib (found suitable exact version "3.10.13") 
CMake Warning (dev) at builds/macosx/_deps/nlopt-src/cmake/FindNumPy.cmake:45 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  builds/macosx/_deps/nlopt-src/CMakeLists.txt:301 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find NumPy (missing: NUMPY_INCLUDE_DIRS) 
error: ("/opt/homebrew/opt/pkg-config/bin/pkg-config" "--variable=prefix" "guile-3.0") exited with non-zero error code 127
error: ("/opt/homebrew/opt/pkg-config/bin/pkg-config" "--variable=sitedir" "guile-3.0") exited with non-zero error code 127
error: ("/opt/homebrew/opt/pkg-config/bin/pkg-config" "--variable=extensiondir" "guile-3.0") exited with non-zero error code 127
-- Could NOT find Guile (missing: GUILE_ROOT_DIR) (found version "3.0.9")
-- Could NOT find SWIG (missing: SWIG_EXECUTABLE SWIG_DIR) 
-- Could NOT find Octave (missing: OCTAVE_EXECUTABLE OCTAVE_ROOT_DIR OCTAVE_INCLUDE_DIRS OCTAVE_LIBRARIES) 
-- Could NOT find Matlab (missing: Matlab_INCLUDE_DIRS MX_LIBRARY MAIN_PROGRAM) (found version "NOTFOUND")
CMake Warning at /opt/homebrew/Cellar/cmake/3.27.7/share/cmake/Modules/CPack.cmake:507 (message):
  CPack.cmake has already been included!!
Call Stack (most recent call first):
  builds/macosx/_deps/nlopt-src/CMakeLists.txt:353 (include)


-- Found exiv2 0.27.6
-- Found Threads: TRUE  
-- Found Gettext: /opt/homebrew/bin/msgmerge (found version "0.22.3") 
-- Found xgettext: /opt/homebrew/bin/xgettext
-- Found lupdate: /Users/aw/Qt/6.5.3/macos/bin/lupdate
-- Found lconvert: /Users/aw/Qt/6.5.3/macos/bin/lconvert
-- Configuring done (23.5s)
CMake Warning (dev) in builds/macosx/_deps/nlopt-src/CMakeLists.txt:
  Policy CMP0071 is not set: Let AUTOMOC and AUTOUIC process GENERATED files.
  Run "cmake --help-policy CMP0071" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  For compatibility, CMake is excluding the GENERATED source file(s):

    "/Users/aw/stellarium/builds/macosx/_deps/nlopt-build/nlopt.hpp"

  from processing by AUTOMOC and AUTOUIC.  If any of the files should be
  processed, set CMP0071 to NEW.  If any of the files should not be
  processed, explicitly exclude them by setting the source file property
  SKIP_AUTOGEN:

    set_property(SOURCE file.h PROPERTY SKIP_AUTOGEN ON)

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done (0.6s)
-- Build files have been written to: /Users/aw/stellarium/builds/macosx
```

Errors:
```
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.cc:8:
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.hpp:10:
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/data_types.hpp:11:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/string:576:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/string_view:1025:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/algorithm:1921:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/memory:898:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:308:9: warning: destructor called on non-final '(anonymous namespace)::ProblemInternal' that has virtual functions but non-virtual destructor [-Wdelete-non-abstract-non-virtual-dtor]
        __get_elem()->~_Tp();
        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:274:14: note: in instantiation of member function 'std::__shared_ptr_emplace<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>>::__on_zero_shared' requested here
    explicit __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:995:55: note: in instantiation of function template specialization 'std::__shared_ptr_emplace<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>>::__shared_ptr_emplace<const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1004:19: note: in instantiation of function template specialization 'std::allocate_shared<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>, const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &, void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.cc:87:19: note: in instantiation of function template specialization 'std::make_shared<(anonymous namespace)::ProblemInternal, const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &, void>' requested here
  mProblem = std::make_shared<ProblemInternal>(functions, leftBound, rightBound);
                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:308:24: note: qualify call to silence this warning
        __get_elem()->~_Tp();
                       ^
1 warning generated.
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.cc:8:
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.hpp:10:
In file included from /Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/data_types.hpp:11:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/string:576:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/string_view:1025:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/algorithm:1921:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/memory:898:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:308:9: warning: destructor called on non-final '(anonymous namespace)::ProblemInternal' that has virtual functions but non-virtual destructor [-Wdelete-non-abstract-non-virtual-dtor]
        __get_elem()->~_Tp();
        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:274:14: note: in instantiation of member function 'std::__shared_ptr_emplace<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>>::__on_zero_shared' requested here
    explicit __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:995:55: note: in instantiation of function template specialization 'std::__shared_ptr_emplace<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>>::__shared_ptr_emplace<const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1004:19: note: in instantiation of function template specialization 'std::allocate_shared<(anonymous namespace)::ProblemInternal, std::allocator<(anonymous namespace)::ProblemInternal>, const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &, void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/Users/aw/stellarium/builds/macosx/_deps/nlopt-src/src/algs/ags/solver.cc:87:19: note: in instantiation of function template specialization 'std::make_shared<(anonymous namespace)::ProblemInternal, const std::vector<std::function<double (const double *)>> &, const std::vector<double> &, const std::vector<double> &, void>' requested here
  mProblem = std::make_shared<ProblemInternal>(functions, leftBound, rightBound);
                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/shared_ptr.h:308:24: note: qualify call to silence this warning
        __get_elem()->~_Tp();
                       ^
[ 75%] Building CXX object plugins/TelescopeControl/src/NexStar/CMakeFiles/TelescopeControl_NexStar.dir/TelescopeClientDirectNexStar.cpp.o
[ 75%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxsimpleooxmlfile.cpp.o
1 warning generated.
[ 75%] Building CXX object _deps/nlopt-build/CMakeFiles/nlopt.dir/src/algs/ags/local_optimizer.cc.o
[ 75%] Building CXX object _deps/nlopt-build/CMakeFiles/nlopt.dir/src/algs/ags/ags.cc.o
[ 76%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxzipreader.cpp.o
[ 76%] Linking CXX static library libnlopt.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: libnlopt.a(mocs_compilation.cpp.o) has no symbols
[ 76%] Built target nlopt
[ 77%] Building CXX object plugins/TelescopeControl/src/Rts2/CMakeFiles/TelescopeControl_Rts2.dir/TelescopeControl_Rts2_autogen/mocs_compilation.cpp.o
[ 77%] Linking CXX static library libTelescopeControl_NexStar.a
[ 77%] Built target TelescopeControl_NexStar
[ 77%] Automatic MOC and UIC for target LensDistortionEstimator-static
AutoUic: /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/lensDistortionEstimatorDialog.ui: Warning: Z-order assignment: 'TitleBar' is not a valid widget.
[ 77%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxcell.cpp.o
[ 77%] Built target LensDistortionEstimator-static_autogen
[ 77%] Building CXX object _deps/showmysky-qt6-build/ShowMySky/CMakeFiles/ShowMySky.dir/ShowMySky_autogen/mocs_compilation.cpp.o
[ 77%] Building CXX object _deps/showmysky-qt6-build/ShowMySky/CMakeFiles/ShowMySky.dir/api/AtmosphereRenderer.cpp.o
[ 77%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxchartsheet.cpp.o
[ 77%] Building CXX object plugins/TelescopeControl/src/Rts2/CMakeFiles/TelescopeControl_Rts2.dir/TelescopeClientJsonRts2.cpp.o
[ 77%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxdocpropsapp.cpp.o
[ 77%] Building CXX object _deps/showmysky-qt6-build/ShowMySky/CMakeFiles/ShowMySky.dir/AtmosphereRenderer.cpp.o
[ 77%] Linking CXX static library libTelescopeControl_Rts2.a
[ 77%] Built target TelescopeControl_Rts2
[ 77%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxmediafile.cpp.o
[ 78%] Building CXX object _deps/showmysky-qt6-build/ShowMySky/CMakeFiles/ShowMySky.dir/util.cpp.o
[ 78%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxstyles.cpp.o
[ 78%] Automatic MOC and UIC for target TelescopeControl_INDI
[ 78%] Built target TelescopeControl_INDI_autogen
[ 78%] Generating qrc_LensDistortionEstimator.cpp
[ 78%] Building CXX object plugins/LensDistortionEstimator/src/CMakeFiles/LensDistortionEstimator-static.dir/LensDistortionEstimator-static_autogen/mocs_compilation.cpp.o
[ 78%] Linking CXX shared library libShowMySky-Qt6.dylib
[ 78%] Built target ShowMySky
[ 78%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxzipwriter.cpp.o
[ 78%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxcellformula.cpp.o
[ 78%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxcolor.cpp.o
[ 78%] Building CXX object plugins/LensDistortionEstimator/src/CMakeFiles/LensDistortionEstimator-static.dir/LensDistortionEstimator.cpp.o
[ 79%] Generating ui_TelescopeClientINDIWidget.h
[ 79%] Generating ui_INDIControlWidget.h
[ 79%] Building CXX object plugins/TelescopeControl/src/INDI/CMakeFiles/TelescopeControl_INDI.dir/TelescopeControl_INDI_autogen/mocs_compilation.cpp.o
[ 79%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxdocpropscore.cpp.o
[ 79%] Building CXX object plugins/TelescopeControl/src/INDI/CMakeFiles/TelescopeControl_INDI.dir/INDIConnection.cpp.o
[ 79%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxnumformatparser.cpp.o
[ 80%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxtheme.cpp.o
[ 81%] Building CXX object plugins/LensDistortionEstimator/src/CMakeFiles/LensDistortionEstimator-static.dir/gui/LensDistortionEstimatorDialog.cpp.o
[ 81%] Building CXX object plugins/TelescopeControl/src/INDI/CMakeFiles/TelescopeControl_INDI.dir/TelescopeClientINDI.cpp.o
[ 81%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxcelllocation.cpp.o
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:54:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<Value> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:256:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<DataValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:338:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<StringValueBase> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:420:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<StringValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:452:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<AsciiValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:540:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<CommentValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:627:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<XmpValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:718:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<XmpTextValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:800:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<XmpArrayValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:893:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<LangAltValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:981:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<DateValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:1082:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<TimeValue> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
In file included from /opt/homebrew/include/exiv2/metadatum.hpp:27:
/opt/homebrew/include/exiv2/value.hpp:1238:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<ValueType<T> > AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
In file included from /opt/homebrew/include/exiv2/datasets.hpp:33:
/opt/homebrew/include/exiv2/metadatum.hpp:47:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<Key> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:26:
/opt/homebrew/include/exiv2/datasets.hpp:278:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<IptcKey> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:27:
/opt/homebrew/include/exiv2/basicio.hpp:58:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<BasicIo> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:27:
/opt/homebrew/include/exiv2/basicio.hpp:524:9: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        std::auto_ptr<Impl> p_;
        ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:27:
/opt/homebrew/include/exiv2/basicio.hpp:724:9: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        std::auto_ptr<Impl> p_;
        ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
In file included from /Users/aw/stellarium/plugins/LensDistortionEstimator/src/gui/LensDistortionEstimatorDialog.cpp:42:
In file included from /opt/homebrew/include/exiv2/exiv2.hpp:28:
In file included from /opt/homebrew/include/exiv2/bmffimage.hpp:27:
In file included from /opt/homebrew/include/exiv2/image.hpp:28:
In file included from /opt/homebrew/include/exiv2/exif.hpp:35:
/opt/homebrew/include/exiv2/tags.hpp:143:17: error: no template named 'auto_ptr' in namespace 'std'; did you mean simply 'auto_ptr'?
        typedef std::auto_ptr<ExifKey> AutoPtr;
                ^~~~~
/opt/homebrew/include/exiv2/config.h:105:3: note: 'auto_ptr' declared here
  using auto_ptr = std::unique_ptr<T>;
  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [plugins/LensDistortionEstimator/src/CMakeFiles/LensDistortionEstimator-static.dir/gui/LensDistortionEstimatorDialog.cpp.o] Error 1
make[1]: *** [plugins/LensDistortionEstimator/src/CMakeFiles/LensDistortionEstimator-static.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 81%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxconditionalformatting.cpp.o
[ 81%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxdocument.cpp.o
[ 81%] Building CXX object plugins/TelescopeControl/src/INDI/CMakeFiles/TelescopeControl_INDI.dir/TelescopeClientINDIWidget.cpp.o
[ 81%] Building CXX object plugins/TelescopeControl/src/INDI/CMakeFiles/TelescopeControl_INDI.dir/INDIControlWidget.cpp.o
[ 81%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxrelationships.cpp.o
[ 81%] Building CXX object _deps/qxlsxqt6-build/CMakeFiles/QXlsx.dir/source/xlsxutility.cpp.o
[ 81%] Linking CXX static library libTelescopeControl_INDI.a
[ 81%] Built target TelescopeControl_INDI
[ 81%] Linking CXX static library libQXlsxQt6.a
[ 81%] Built target QXlsx
make: *** [all] Error 2

```

macOS 14.1.1 / M1

alex-w commented:
OK, remove `exiv2` from `homebrew` (errors are expected due architecture) and try it again...

10110111 commented:
> ```
> /opt/homebrew/include/exiv2/value.hpp:54:17: error: no template named 'auto_ptr' in namespace > 'std'; did you mean simply 'auto_ptr'?
>         typedef std::auto_ptr<Value> AutoPtr;
>                 ^~~~~
> ```


See the recipe I added in the CI workflow to learn how this should be treated.

alex-w commented:
Is it not possible add exiv2 via CPM?

10110111 commented:
> Is it not possible add exiv2 via CPM?

It has several dependencies that I don't know how to handle via CPM. You can try yourself ;)

10110111 commented:
So now all x86/amd64 builds have exiv2 available in the CI, only ARM64 remains without it. I think this is good enough for a start. Any reviews?..

ElonVampire commented:
收到！
